### DISCUSSION TITLE
How do I actively test my code while developing

### DISCUSSION THREAD

**Discussion by awalvie** [kubernetes-sigs/kubebuilder#3768] [Support] [answered]
Currently following the `CronJob` tutorial from the kubebuilder book. I was stepping through https://book.kubebuilder.io/cronjob-tutorial/controller-implementation and realized that if I was on my own, writing the reconciliation logic from scratch I would have no idea what functions/structs/methods to make use of.

To give an analogy -- say I'm working on the backend of a webserver. I create a new route that gets the input from a form and writes it to a database. Here's how I would go about implementing said route:

* Create a route that does nothing
* Access the route -- ensure its hitting my function
* Print out the contents of the object I'm receiving in the function to figure out what structs/objects my incoming data is
* Write the unmarshal logic
* Access the route -- verify unmarshalling logic by printing out my unmarshaled struct
* Write the logic to push to database
...

you get the idea. There's an active feedback loop in the above workflow so that there's no guesswork involved. In the same way say if I was implementing the reconciliation logic on my own here's how I would start:

* Run the controller(?)
* Try to hit the `Reconcile()` function and print out the contexts of the request
* Write the logic for getting all the `CronJob`s
* Reload controller -- see the output

so on and so forth. Are there any docs that highlight what this process is like, if not, any ideas on how I could do this?

**Comment by camilamacedo86** [comment, marked as answer]
Hi @awalvie,

### Overview

To validate your controller logic, you can use [ENVTEST](https://book.kubebuilder.io/reference/envtest.html). Check out the examples under the controllers to create the suite and the test for each controller [here](https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata/project-v4/internal/controller). Note that you should have a test for each controller. 

Additionally, you can perform end-to-end (e2e) tests to check your solution, with an example available under the testdata [here](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4/test/e2e/e2e_test.go). When creating a project with the Kubebuilder CLI, an initial scaffold is provided in `test/e2e` for your e2e tests. You can also refer to `suite_test.go` for using envtest.

If you use the **Deploy Image** plugin (more information [here](https://book.kubebuilder.io/plugins/deploy-image-plugin-v1-alpha)), it generates the scaffold to deploy and manage an image, including an implemented test. See this example [here](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-deploy-image/internal/controller/memcached_controller_test.go).

The Kubebuilder docs provide a section on writing tests with examples [here](https://book.kubebuilder.io/cronjob-tutorial/writing-tests).

Remember that following up the Operator Pattern, we declare the desired state and loop (reconcile) until it is ensured. It's recommended to re-trigger the reconciliation based on events raised, like updates, deletions, or creations of a Custom Resource that you are watching. Therefore, you can use kubectl, for example, to apply the CR and check the logs to know what is happening. 

💡 Check the code generated by the Deploy Image plugin to understand Conditional Status and best practices. Review the Kubernetes API conventions [here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status). The implementation can be seen in the code generated by the Deploy Image plugin [here](https://github.com/kubernetes-sigs/kubebuilder/blob/919502e3f36e68f870ca2d1c59714fc18ce60dbf/testdata/project-v4-with-deploy-image/api/v1alpha1/busybox_types.go#L26-L52).

Also, use the e2e tests implemented in Kubebuilder as a reference, available [here](https://github.com/kubernetes-sigs/kubebuilder/tree/master/test/e2e).

### Iterative Development Process

For an iterative development process in Kubebuilder:

- **Initial Setup**: Start with a basic Reconcile function that logs its invocation.
  
  ```go
  func (r *CronJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
      log.Printf("Reconciling CronJob: %s", req.NamespacedName)
      // Your code here
      return ctrl.Result{}, nil
  }
  ```

### Testing and Logging
- **Run the controller** and trigger the `Reconcile` function by making changes to your Kubernetes objects (like CronJob).
- Use **logging** to understand the context and request data you receive in the `Reconcile` function.

### Iterative Development
- **Implement small logic parts**, such as listing all CronJobs.
- **Verify your code** with logging.
- **Gradually add more logic**, for instance, processing each CronJob, handling errors, and updating the status.

### Feedback Loop
- **Test each new logic addition** by re-running the controller.
- Use **Kubernetes tools** like `kubectl` to modify objects and observe your controller's reaction.
- **Adjust your code** based on the output and behavior you observe.

### Additional Tips
- For **ENVTEST tests**, run `make test` to execute them. You can configure it to run against an existing cluster and use tools like **kind** for local testing.
- Consider **setting up GitHub Actions** with e2e tests and the `make test` target to ensure changes are validated when pushing PRs.
- **Debugging tests** can be done with tools like **GolangIDEA**.

_I am closing this as sorted out, but feel free to reopen it if needed, or open new discussion topics for more specific questions after reviewing all the content and examples shared._

**Reply by awalvie** [reply]
Thank you so much for such a detailed response. Really appreciate it. Cheers!
